### PR TITLE
Signedness value

### PR DESCRIPTION
**Pull request by cmackie**
This branch integrates the Value Checker into the Signedness Checker and uses it to add new introduction rules for identifiers.

### CODE REVIEW

File: checker/src/org/checkerframework/checker/signedness/SignednessAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -109,6 +121,14 @@ protected TreeAnnotator createTreeAnnotator() {
     }
 
     /**
+     * Provides a way to query the Constant Value Checker, which computes the values of expressions
+     * known at compile time (constant propagation and folding).
+     */
+    ValueAnnotatedTypeFactory getValueAnnotatedTypeFactory() {
```
Comment (mernst) [inline checker/src/org/checkerframework/checker/signedness/SignednessAnnotatedTypeFactory.java]: How about caching this value for efficiency?  Also, I think this method should be private.

File: checker/src/org/checkerframework/checker/signedness/SignednessAnnotatedTypeFactory.java [hunk 2]
```diff
@@ -154,5 +174,42 @@ public Void visitCompoundAssignment(CompoundAssignmentTree tree, AnnotatedTypeMi
             annotateBoolean(type);
             return null;
         }
+
+        @Override
+        public Void visitIdentifier(IdentifierTree tree, AnnotatedTypeMirror type) {
+            TypeMirror javaType = type.getUnderlyingType();
+            ValueAnnotatedTypeFactory valFact = getValueAnnotatedTypeFactory();
```
Comment (mernst) [inline checker/src/org/checkerframework/checker/signedness/SignednessAnnotatedTypeFactory.java]: For efficiency, should these two lines be executed only if javaType is a primitive?

File: checker/src/org/checkerframework/checker/signedness/SignednessAnnotatedTypeFactory.java [hunk 3]
```diff
@@ -154,5 +174,42 @@ public Void visitCompoundAssignment(CompoundAssignmentTree tree, AnnotatedTypeMi
             annotateBoolean(type);
             return null;
         }
+
+        @Override
+        public Void visitIdentifier(IdentifierTree tree, AnnotatedTypeMirror type) {
```
Comment (mernst) [inline checker/src/org/checkerframework/checker/signedness/SignednessAnnotatedTypeFactory.java]: Please write a one-line `//` comment to say what this overriding implementation does.

File: checker/src/org/checkerframework/checker/signedness/SignednessAnnotatedTypeFactory.java [hunk 4]
```diff
@@ -154,5 +174,42 @@ public Void visitCompoundAssignment(CompoundAssignmentTree tree, AnnotatedTypeMi
...
+            TypeMirror javaType = type.getUnderlyingType();
+            ValueAnnotatedTypeFactory valFact = getValueAnnotatedTypeFactory();
+            Range treeRange = IndexUtil.getPossibleValues(valFact.getAnnotatedType(tree), valFact);
+
+            if (treeRange != null) {
+                switch (javaType.getKind()) {
+                    case BYTE:
```
Comment (mernst) [inline checker/src/org/checkerframework/checker/signedness/SignednessAnnotatedTypeFactory.java]: This only handles primitives, but the Value Checker handles wrappers such as `Integer`.  Please support them or write a comment explaining why they are not supported.

File: checker/src/org/checkerframework/checker/signedness/SignednessAnnotatedTypeFactory.java (hunk 1 above)
Comment (smillst) [inline checker/src/org/checkerframework/checker/signedness/SignednessAnnotatedTypeFactory.java]: The annotated type factory of a subchecker should not be cached.  See [the documentation of getTypeFactoryOfSubchecker](https://checkerframework.org/api/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.html#getTypeFactoryOfSubchecker-java.lang.Class-).

**Comment by mernst** [comment]
Chris, can you answer my question about why the rules are only applied to identifiers?
Thanks!

**Comment by cmackie** [comment]
The new rules can absolutely be applied to expressions, and should be. I ran out of time to implement those rules for expressions, but I figured the rules as applied to literals could be pushed now since they are sound and still pretty helpful.

**Comment by mernst** [comment]
If you won't have time to expand the rules beyond expressions, could you document in the code that doing so is desirable, and also add an issue so that someone fixes this in the future?

**Comment by mernst** [comment]
A ping on the documentation and new issue I suggested.  Then we can close this issue, in advance of the August 1 release.

**Comment by smillst** [comment]
The new issue is https://github.com/typetools/checker-framework/issues/1430.

**Pull request merged**